flood_api/utils/geospatial_operations.py

```diff
@@ -0,0 +1,192 @@
+                    bottom_right, bottom_left])
+
+def get_data_for_point(
+    latitude: float, 
+    longitude: float, 
+    gdf: gpd.GeoDataFrame,
+    include_neighbors: bool = False, 
+    date_range: tuple[date, date] = None
+) -> tuple[gpd.GeoDataFrame, gpd.GeoDataFrame]:
+    """
+    Given a latitude and longitude, return the data for the grid cell
+    it falls into. Optionally, include the data for neighboring cells.
+    A date range can also be provided to filter the data, yielding only
+    the data within the date range (inclusive). If no date range is provided,
+    data for all dates will be returned.
+    
+    The function identifies the grid cell based on the following logic:
+    - If the provided point is on the boundary between two cells, the function 
+      will assign the point to the cell to its east (right, for longitude) or 
+      north (above, for latitude).
+    - For example, for a grid_size of 0.05:
+        * An input latitude of -5.8 and longitude of 37.75 would result in a 
+            bounding box from latitude -5.8 to -5.75 and longitude 37.75 to 37.8.
+        * An input latitude of -5.81 and longitude of 37.7501 would result in a 
+            bounding box from latitude -5.85 to -5.8 and longitude 37.75 to 37.8.
+    
+    Parameters:
+    - latitude (float): The latitude of the point.
+    - longitude (float): The longitude of the point.
+    - gdf (GeoDataFrame): The GeoDataFrame to query.
+    - include_neighbors (bool): Whether to include neighboring cells. Defaults to False.
+    - date_range (tuple, optional): The date range to query (inclusive). Defaults to None.
+
+    Returns:
+    tuple: The primary cell and neighbors as GeoDataFrames.
+    """
+    # Get the bounds for the queryed cell
+    cell_bounds = get_grid_cell_bounds(
+        latitude=latitude, 
+        longitude=longitude, 
+        grid_size=GLOFAS_RESOLUTION, 
+        precision=GLOFAS_PRECISION
+    )
+
+    # Get the deflated geometry to find primary cell
+    reduced_geometry = create_polygon_from_bounds(
+        *cell_bounds, 
+        buffer=-GLOFAS_RESOLUTION/2,
+        precision=GLOFAS_PRECISION
+    )
+    
+    neighbors_only_df = gpd.GeoDataFrame(geometry=[])
+
+    if date_range is not None:
+        # Filter the dataframe for the date range
+        # Since 'valid_time' represents the end of the 24 hour period,
+        # we need to add one day to the end date to get the correct results.
+        # For example:
+        # If the start date is 2021-01-01 and the end date is 2021-01-03,
+        # the query will return data for which valid_time is equal to 
+        # 2021-01-02, 2021-01-03, and 2021-01-04, since valid_time represents
+        # the end of the 24 hour period.
+        start_date, end_date = [Timestamp(date) for date in date_range]
+        if end_date != Timestamp(date.max):
+            end_date += timedelta(days=1)
+        gdf = gdf[(gdf['valid_time'] > start_date) & (gdf['valid_time'] <= end_date)]
```

Comment:
I would try to unify the input-dates and the output dates some way or another

Reply:
Ok. We can change `valid_time` to `valid_for` and let it represent the day for which the forecast is valid. So if the `valid_time` field used to be `'2023-11-30 00:00:00'`, `valid_for` will be `'2023-11-29'`. This will introduce consistency with the queried `start-date` and `end-date` parameters. We can drop the time and keep these as dates (fastparquet seems to read the dates as `pd.Timestamp` by default, so this will have to be overriden).